### PR TITLE
Fix IIR -> Optimizer dependency

### PR DESCRIPTION
**Pull request by twicki**
## Technical Description

We remove all the backwards dependencies from the IIR to the Optimizer. This is part of cleaning up all unwanted (circular) dependencies between modules

####  Enhances

Now IIR is independent of Optimizer

## Dependencies

This PR is independent

### CODE REVIEW

**Comment by twicki** [comment]
launch pipeline

File: src/dawn/IIR/StencilInstantiation.h [hunk 1]
```diff
@@ -111,23 +111,23 @@ class StencilInstantiation : NonCopyable {
   /// @param expr       Expression to be renamed (usually the left- or right-hand side of an
   ///                   assignment). Can be `NULL`.
   /// @returns AccessID of the new field
-  int createVersionAndRename(int AccessID, Stencil* stencil, int stageIndex, int stmtIndex,
-                             std::shared_ptr<iir::Expr>& expr, RenameDirection dir);
+  // int createVersionAndRename(int AccessID, Stencil* stencil, int stageIndex, int stmtIndex,
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/IIR/StencilInstantiation.h]: done

File: src/dawn/IIR/StencilInstantiation.h [hunk 2]
```diff
@@ -111,23 +111,23 @@ class StencilInstantiation : NonCopyable {
...
+  //                            std::shared_ptr<Expr>& expr, RenameDirection dir);
 
   /// @brief Rename all occurences of field `oldAccessID` to `newAccessID`
-  void renameAllOccurrences(Stencil* stencil, int oldAccessID, int newAccessID);
+  // void renameAllOccurrences(Stencil* stencil, int oldAccessID, int newAccessID);
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/IIR/StencilInstantiation.h]: done, sorry

File: src/dawn/Optimizer/PassFieldVersioning.cpp [hunk 3]
```diff
@@ -276,9 +277,12 @@ PassFieldVersioning::RCKind PassFieldVersioning::fixRaceCondition(
 
   // Create a new multi-versioned field and rename all occurences
   for(int oldAccessID : renameCandiates) {
-    int newAccessID = instantiation->createVersionAndRename(oldAccessID, &stencil, stageIdx, index,
-                                                            assignment->getRight(),
-                                                            iir::StencilInstantiation::RD_Above);
+    int newAccessID = createVersionAndRename(instantiation.get(), oldAccessID, &stencil, stageIdx,
+                                             index, assignment->getRight(), RenameDirection::Above);
+    // int newAccessID = instantiation->createVersionAndRename(oldAccessID, &stencil, stageIdx,
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/PassFieldVersioning.cpp]: done

File: src/dawn/Optimizer/PassSSA.cpp [hunk 4]
```diff
@@ -73,10 +74,14 @@ bool PassSSA::run(const std::shared_ptr<iir::StencilInstantiation>& stencilInsta
           tochedAccessIDs.insert(AccessID);
         }
 
-        for(int AccessID : AccessIDsToRename)
-          tochedAccessIDs.insert(stencilInstantiation->createVersionAndRename(
-              AccessID, &stencil, stageIdx, stmtIdx, assignment->getLeft(),
-              iir::StencilInstantiation::RD_Below));
+        for(int AccessID : AccessIDsToRename) {
+          // tochedAccessIDs.insert(stencilInstantiation->createVersionAndRename(
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/PassSSA.cpp]: done

File: src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp [hunk 5]
```diff
@@ -606,9 +609,12 @@ bool PassTemporaryToStencilFunction::run(
       if(metadata.isAccessType(iir::FieldAccessType::FAT_GlobalVariable, varID))
         continue;
 
-      stencilInstantiation->promoteLocalVariableToTemporaryField(
-          stencilPtr.get(), varID, stencilPtr->getLifetime(varID),
-          iir::TemporaryScope::TS_StencilTemporary);
+      // stencilInstantiation->promoteLocalVariableToTemporaryField(
+      //     stencilPtr.get(), varID, stencilPtr->getLifetime(varID),
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp]: done

File: src/dawn/Optimizer/TemporaryHandeling.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===/
+
+#include "dawn/Optimizer/TemporaryHandeling.h"
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/TemporaryHandeling.cpp]: done

File: src/dawn/IIR/StencilFunctionInstantiation.cpp [hunk 7]
```diff
@@ -658,7 +633,7 @@ void StencilFunctionInstantiation::dump() const {
 
   const auto& statements = getAST()->getRoot()->getStatements();
   for(std::size_t i = 0; i < statements.size(); ++i) {
-    std::cout << "\e[1m" << iir::ASTStringifier::toString(statements[i], 2 * DAWN_PRINT_INDENT)
+    std::cout << "\e[1m" << ast::ASTStringifier::toString(statements[i], 2 * DAWN_PRINT_INDENT)
```
Comment (Stagno) [inline src/dawn/IIR/StencilFunctionInstantiation.cpp]: iir

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 8]
```diff
@@ -20,7 +20,8 @@
 #include "dawn/Optimizer/OptimizerContext.h"
 #include "dawn/Optimizer/PassDataLocalityMetric.h"
 #include "dawn/Optimizer/PassSetCaches.h"
-#include "dawn/IIR/ASTExpr.h"
+#include "dawn/Optimizer/Renaming.h"
+#include "dawn/SIR/ASTExpr.h"
```
Comment (Stagno) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: IIR/ASTExpr.h

File: src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp [hunk 9]
```diff
@@ -24,6 +24,9 @@
 #include "dawn/Optimizer/AccessComputation.h"
 #include "dawn/Optimizer/OptimizerContext.h"
 #include "dawn/Optimizer/StatementMapper.h"
+#include "dawn/Optimizer/TemporaryHandling.h"
+#include "dawn/SIR/AST.h"
+#include "dawn/SIR/ASTVisitor.h"
```
Comment (Stagno) [inline src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp]: IIR/ASTVisitor.h

File: src/dawn/Optimizer/PassTemporaryType.cpp [hunk 10]
```diff
@@ -19,7 +19,8 @@
 #include "dawn/IIR/Stencil.h"
 #include "dawn/IIR/StencilInstantiation.h"
 #include "dawn/Optimizer/OptimizerContext.h"
-#include "dawn/IIR/ASTVisitor.h"
+#include "dawn/Optimizer/TemporaryHandling.h"
+#include "dawn/SIR/ASTVisitor.h"
```
Comment (Stagno) [inline src/dawn/Optimizer/PassTemporaryType.cpp]: IIR/ASTVisitor

File: src/dawn/IIR/DependencyGraphAccesses.cpp [hunk 11]
```diff
@@ -535,5 +603,20 @@ void DependencyGraphAccesses::toJSON(const std::string& file, DiagnosticsEngine&
   ofs.close();
 }
 
+bool DependencyGraphAccesses::exceedsMaxBoundaryPoints(int maxHorizontalBoundaryExtent) {
+  // auto nodeExtentsPtr = computeBoundaryExtents(this);
+  std::unordered_map<std::size_t, Extents> extentMap = *computeBoundaryExtents(this);
```
[the remark this answers is not in the document]
Comment (lukasm91) [reply src/dawn/IIR/DependencyGraphAccesses.cpp]: It involves a deep copy, but your proposal leaks memory. computeBoundaryExtent should not return a unique_ptr.

File: src/dawn/Optimizer/OptimizerContext.cpp [hunk 12]
```diff
@@ -78,11 +78,14 @@ class StencilDescStatementMapper : public iir::ASTVisitor {
   /// should insert a call to the gridtools stencil here
   std::shared_ptr<iir::Stmt> stencilDescReplacement_;
 
+  OptimizerContext* context_;
```
Comment (lukasm91) [inline src/dawn/Optimizer/OptimizerContext.cpp]: could it be a reference? (can context be nullptr or go out of scope?)

File: src/dawn/Optimizer/ReorderStrategy.h [hunk 13]
```diff
@@ -15,10 +15,11 @@
 #ifndef DAWN_OPTIMIZER_REORDERSTRATEGY_H
 #define DAWN_OPTIMIZER_REORDERSTRATEGY_H
 
+#include "dawn/Optimizer/OptimizerContext.h"
 #include <memory>
 
 namespace dawn {
-
+class OptimizerContext;
```
Comment (lukasm91) [inline src/dawn/Optimizer/ReorderStrategy.h]: I would not forward declare it if you include the file

File: src/dawn/Serialization/IIRSerializer.cpp [hunk 14]
```diff
@@ -798,7 +798,7 @@ IIRSerializer::deserialize(const std::string& file, OptimizerContext* context,
 
   std::string str((std::istreambuf_iterator<char>(ifs)), std::istreambuf_iterator<char>());
   std::shared_ptr<iir::StencilInstantiation> returnvalue =
-      std::make_shared<iir::StencilInstantiation>(context);
+      std::make_shared<iir::StencilInstantiation>();
```
Comment (lukasm91) [inline src/dawn/Serialization/IIRSerializer.cpp]: we need to make sure that we all use the same clang-format version :-)

Comment (twicki) [reply src/dawn/Serialization/IIRSerializer.cpp]: the  context actually got removed here, something did change :)

File: src/dawn/Optimizer/ReorderStrategy.h (hunk 13 above)
Comment (twicki) [reply src/dawn/Optimizer/ReorderStrategy.h]: done

File: src/dawn/Optimizer/CreateVersionAndRename.cpp [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+#include "dawn/Optimizer/CreateVersionAndRename.h"
+#include "dawn/IIR/StencilInstantiation.h"
+#include "dawn/Optimizer/AccessComputation.h"
+#include "dawn/Optimizer/Renaming.h"
+
+namespace dawn {
+
+int createVersionAndRename(iir::StencilInstantiation* instantiation, int AccessID,
+                           iir::Stencil* stencil, int curStageIdx, int curStmtIdx,
+                           std::shared_ptr<ast::Expr>& expr, RenameDirection dir) {
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/CreateVersionAndRename.cpp]: done

File: src/dawn/Optimizer/CreateVersionAndRename.h [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+/// @param AccessID   AccessID of the field for which a new version will be created
+/// @param stencil    Current stencil
+/// @param stageIdx   **Linear** index of the stage in the stencil
+/// @param stmtIdx    Index of the statement inside the stage
+/// @param expr       Expression to be renamed (usually the left- or right-hand side of an
+///                   assignment). Can be `NULL`.
+/// @returns AccessID of the new field
+int createVersionAndRename(iir::StencilInstantiation* instantiation, int AccessID,
+                           iir::Stencil* stencil, int curStageIdx, int curStmtIdx,
+                           std::shared_ptr<ast::Expr>& expr, RenameDirection dir);
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/CreateVersionAndRename.h]: done

File: src/dawn/Optimizer/Renaming.h [hunk 17]
```diff
@@ -50,7 +52,7 @@ void renameAccessIDInStmts(
     iir::StencilFunctionInstantiation* instantiation, int oldAccessID, int newAccessID,
     ArrayRef<std::unique_ptr<iir::StatementAccessesPair>> statementAccessesPairs);
 void renameAccessIDInExpr(iir::StencilInstantiation* instantiation, int oldAccessID,
-                          int newAccessID, std::shared_ptr<iir::Expr>& expr);
+                          int newAccessID, std::shared_ptr<ast::Expr>& expr);
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/Renaming.h]: done

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 8 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: done

File: src/dawn/IIR/StencilFunctionInstantiation.cpp (hunk 7 above)
Comment (twicki) [reply src/dawn/IIR/StencilFunctionInstantiation.cpp]: done

File: src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp (hunk 9 above)
Comment (twicki) [reply src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp]: done

File: src/dawn/Optimizer/PassTemporaryType.cpp (hunk 10 above)
Comment (twicki) [reply src/dawn/Optimizer/PassTemporaryType.cpp]: done

File: src/dawn/IIR/DependencyGraphAccesses.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/IIR/DependencyGraphAccesses.cpp]: there is fact no reason for smart pointers here. (The function was just moved, I didn't touch it). I proposed a solution returning a map and we can let RVO take care of it

[the remark this answers is not in the document]
Comment (Stagno) [reply src/dawn/IIR/DependencyGraphAccesses.cpp]: seems reasonable

File: src/dawn/Serialization/IIRSerializer.cpp (hunk 14 above)
Comment (lukasm91) [reply src/dawn/Serialization/IIRSerializer.cpp]: somehow the comment is at the wrong place...it's for line 646.

**Comment by twicki** [comment]
launch pipeline

**Pull request merged**